src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11Util.java

```diff
@@ -46,37 +47,37 @@ public final class P11Util {
 
     // Used by PBE
     static final class KDFData {
-        public enum Operation {ENCRYPTION, AUTHENTICATION, GENERIC}
-        public long kdfMech;
-        public long prfMech;
-        public String keyAlgo;
-        public int keyLen;
-        public Operation op;
+        public final long kdfMech;
+        public final long prfMech;
+        public final String keyAlgo;
+        public final int keyLen;
+        public final CK_ATTRIBUTE encrypt_or_sign_true;
+
         KDFData(long kdfMech, long prfMech, String keyAlgo,
-                int keyLen, Operation op) {
+                int keyLen, CK_ATTRIBUTE encrypt_or_sign_true) {
             this.kdfMech = kdfMech;
             this.prfMech = prfMech;
             this.keyAlgo = keyAlgo;
             this.keyLen = keyLen;
-            this.op = op;
+            this.encrypt_or_sign_true = encrypt_or_sign_true;
         }
 
         public static void addPbkdf2Data(String algo, long kdfMech,
                                          long prfMech) {
             kdfDataMap.put(algo, new KDFData(kdfMech, prfMech,
-                    "Generic", -1, Operation.GENERIC));
+                    algo, -1, CK_ATTRIBUTE.SIGN_TRUE));
         }
 
         public static void addPbkdf2AesData(String algo, long kdfMech,
                                             long prfMech, int keyLen) {
             kdfDataMap.put(algo, new KDFData(kdfMech, prfMech,
-                    "AES", keyLen, Operation.ENCRYPTION));
+                    "AES", keyLen, CK_ATTRIBUTE.ENCRYPT_TRUE));
```

Comment:
Why are AES keys different? What are the consequences of that?

Reply:
We have redesigned the map, so this comment does not longer apply. From now on, "AES" in this particular case will be an available "keyUsage" for AES-based PBE key types. In fact, all PBE keys (except the ones of PBKDF2 type) have "keyUsage" information. All PBE keys (including those of PBKDF2 type) decide with their own method which are key usages allowed. By usage we mean: can a key of this PBE type be used for the XYZ service algorithm?